packages/flutter/lib/src/painting/image_provider.dart

```diff
@@ -460,11 +477,51 @@ abstract class AssetBundleImageProvider extends ImageProvider<AssetBundleImageKe
   ///
   /// This function is used by [load].
   @protected
-  Future<ui.Codec> _loadAsync(AssetBundleImageKey key) async {
+  Future<ui.Codec> _loadAsync(AssetBundleImageKey key, DecoderCallback decode) async {
     final ByteData data = await key.bundle.load(key.name);
     if (data == null)
       throw 'Unable to read data';
-    return await PaintingBinding.instance.instantiateImageCodec(data.buffer.asUint8List());
+    return await decode(data.buffer.asUint8List());
+  }
+}
+
+class _SizeAwareCacheKey {
+  const _SizeAwareCacheKey(this._providerCacheKey, this._width, this._height);
+
+  final dynamic _providerCacheKey;
+  final int _width;
+  final int _height;
+}
+
+/// Re-sizes the cache of the image provided to the specified size.
+class ResizedImage extends ImageProvider<_SizeAwareCacheKey> {
+  /// Creates an object that re-sizes the cache of the image to the specified size.
+  ///
+  /// The image cached and returned through the [ImageStreamCompleter] will be
+  /// resized from its native resolution.
+  const ResizedImage(this._imageProvider, this._cacheWidth, this._cacheHeight);
+
+  final ImageProvider _imageProvider;
+
+  final int _cacheWidth;
+  final int _cacheHeight;
+
+  @override
+  ImageStreamCompleter load(_SizeAwareCacheKey key, DecoderCallback decode) {
+    if (_cacheWidth == null && _cacheHeight == null) {
+      return _imageProvider.load(key._providerCacheKey, decode);
+    } else {
+      DecoderCallback decodeResize = (Uint8List bytes, {int cacheWidth, int cacheHeight}) {
+        return decode(bytes, cacheWidth: _cacheWidth, cacheHeight: _cacheHeight);
```

Comment:
should we assert that cacheWidth and cacheHeight are null here? e.g. consider what happens if you nest a ResizedImage in another.